### ISSUE TITLE
Redundant calls to search_index.insert_or_update_object() for Documents and Images

### ISSUE THREAD

**Issue by bill-filler** [wagtail/wagtail#4518]
Found a bug? Please fill out the sections below. 👍

### Issue Summary

When uploading a Document or Image, the object is being updated twice in the search backend when it should only be happening once. For Document chooser, the call to form.save() here:
https://github.com/wagtail/wagtail/blob/master/wagtail/documents/views/chooser.py#L113 causes the search signal handler post_save() to update the search index:
https://github.com/wagtail/wagtail/blob/master/wagtail/search/signal_handlers.py#L23.

Then a few lines later in the chooser code, an explicit call is made to search_index.insert_or_update_object():
https://github.com/wagtail/wagtail/blob/master/wagtail/documents/views/chooser.py#L116

This call seems unnecessary as it's already happening as a result of form.save() call triggering the search signal handler and shouldn't need to be called again explicitly

Also happening in documents.py:
https://github.com/wagtail/wagtail/blob/master/wagtail/documents/views/documents.py#L100
https://github.com/wagtail/wagtail/blob/master/wagtail/documents/views/documents.py#L138

Same is true for images:
https://github.com/wagtail/wagtail/blob/master/wagtail/images/views/chooser.py#L133
https://github.com/wagtail/wagtail/blob/master/wagtail/images/views/images.py#L114
https://github.com/wagtail/wagtail/blob/master/wagtail/images/views/images.py#L254

### Steps to Reproduce

1. Enable elasticsearch search backend
2. set a breakpoint or put print statement in search backend add_item() method
https://github.com/wagtail/wagtail/blob/master/wagtail/search/backends/elasticsearch2.py#L797 or 
3. upload a Document via the Document Chooser dialog

Expected Results:
- Should only see one print out in search backend add_item()

Actual Result:
- See two print outs in search backend add_item()

### Technical details

* Python version: 3.6
* Django version: 1.11
* Wagtail version: 1.13.1
* Browser version: Chrome

**Comment by kaedroho** [comment]
It's implemented this way in order for the tags to reindex. This is because Django calls the ``post_save`` signal handler before it saves the tags.

Ideally, we would ignore the ``post_save`` signal for documents or images or find a way to run a signal handler after related objects have been saved.

**Comment by laymonage** [comment]
> Ideally, we would ignore the `post_save` signal for documents or images or find a way to run a signal handler after related objects have been saved.

Wouldn't Django's [`m2m_changed`](https://docs.djangoproject.com/en/stable/ref/signals/#m2m-changed) signal be the appropriate way to do this?

Not sure if the `post_save` signal should be ignored, as I imagine we'd still want to update the index if only direct fields have been changed, while the tags M2M remain unchanged.

If we had handlers for both `post_save` _and_ `m2m_changed`, we would avoid the redundant call when you don't change the tags, as `m2m_changed` won't fire (I think?). Though we'd still end up with redundant calls if you change the tags even if you don't update the direct fields, since `post_save` will be called regardless of change. To avoid that redundant call, the `post_save` handler must know if any direct fields have changed at all, which seems too much magic/complexity for little benefit.

Two changes in the landscape worth noting since this issue was filed:

- After #8803 (Wagtail 4.0), the "redundant" call now lives in `BaseDocumentForm.save()` and `BaseImageForm.save()`: https://github.com/wagtail/wagtail/blob/20259838f089138e1319b51621677358b90c8268/wagtail/documents/forms.py#L67-L68 https://github.com/wagtail/wagtail/blob/20259838f089138e1319b51621677358b90c8268/wagtail/images/forms.py#L69-L70
- As of Wagtail 7.2, the signal handlers now live in https://github.com/wagtail/django-modelsearch/blob/main/modelsearch/signal_handlers.py
  - If we were to proceed with adding an `m2m_changed` handler, we should probably only add it if the model defines a `index.RelatedFields` in `search_fields`.
- The "update index" operation is now done as a background task. This means on non-immediate task backends, the m2m handling is probably not needed, as the index update would likely be deferred until after the m2m was saved in the request-response cycle.
  - We'd still need it if people use the immediate backend, though.

---

For this to proceed, I think I would appreciate if someone with experience can confirm:

- Whether we should add an `m2m_changed` signal.
- If so, whether to add it to `django-modelsearch` for all indexed models that define `index.RelatedFields`, or only do this in Wagtail for Document and Image models.
- If the behaviour stays correct after adding the `m2m_changed` signal and removing the `insert_or_update_object()` calls in both forms' `save()` method.
